### PR TITLE
rafthttp: add leader to transport if peer does not exist

### LINKED ISSUE #3699: Minority failure during cluster configuration change risks deadlock
Adding a node to a degraded cluster and then electing it as the leader causes the addition not to reach the nodes that were down. In particular, starting with a cluster of 3 nodes `a`, `b`, `c` and adding `d` while `a` is not available and then electing `d` as the leader makes `a` unable to participate in the cluster. `a` will remain useless as long as `d` is the leader. This is especially bad in case another node (e.g. `c`) goes down, and the cluster would have to rely on the participation of `a` to make progress. Thus it is possible to have a situation where there is a live majority from both the old cluster (`a`, `b`) and the new cluster (`a`, `b`, `d`) but the cluster is stuck because `a` does not listen to `d` even though `d` is the leader.

I am not sure how to fix this. The invariant from the thesis does not seem applicable since etcd configuration changes apply at a different time and I do not understand the correctness reasoning behind the etcd configuration change algorithm well enough to make changes to it.

A screencast of me reproducing this issue is available at
http://web.mit.edu/andreser/Public/etcd-reconfiguration-deadlock/index.html

### PR DESCRIPTION
**Pull request by heyitsanthony**
cluster integration now supports adding members with stopped nodes, too

Fixes #3699

### CODE REVIEW

**Comment by heyitsanthony** [comment]
not sure if fetching the member list is the best way to go about fixing the problem, but it seems to work /cc @xiang90

**Comment by xiang90** [comment]
@heyitsanthony Another idea is that the peer always attaches its own urls into the http header. So that the follower can always update/add it when it receives anything from another peer.

**Comment by heyitsanthony** [comment]
I like that better. Will give it a try.

**Comment by xiang90** [comment]
@heyitsanthony Thanks.

**Comment by heyitsanthony** [comment]
now sending peers over http headers, ptal /cc @xiang90

File: integration/cluster.go [hunk 1]
```diff
@@ -299,6 +321,11 @@ func (c *cluster) waitLeader(t *testing.T, membs []*member) int {
 	for lead == 0 || !possibleLead[lead] {
 		lead = 0
 		for _, m := range membs {
+			select {
+			case <-m.s.StopNotify():
```
Comment (xiang90) [inline integration/cluster.go]: this looks tricky. i would rather add a stopped bool or something.

File: rafthttp/http.go [hunk 2]
```diff
@@ -271,6 +273,19 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	peers := []string{}
+	for id, p := range h.peerGetter.list() {
+		var urls []string
+		for _, url := range p.urls() {
+			urls = append(urls, url.String())
+		}
+		peers = append(peers, fmt.Sprintf("%x=%s", uint64(id), strings.Join(urls, ",")))
+	}
+	// peerId0=peerURL,...;peerID1=peerURL,...;...
+	w.Header().Set("X-Raft-Peers", strings.Join(peers, ";"))
```
Comment (xiang90) [inline rafthttp/http.go]: I feel we only need to attach the local peer not the entire peer list.

File: rafthttp/stream.go [hunk 3]
```diff
@@ -407,6 +407,18 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
 		}
 		return nil, err
 	case http.StatusOK:
+		if peers := resp.Header.Get("X-Raft-Peers"); peers != "" {
```
Comment (xiang90) [inline rafthttp/stream.go]: parsePeerHander?

File: rafthttp/http.go (hunk 2 above)
Comment (heyitsanthony) [inline rafthttp/http.go]: The follower doesn't know the local peer address, so it wouldn't be able to connect to get the header.

File: integration/cluster.go (hunk 1 above)
Comment (heyitsanthony) [inline integration/cluster.go]: I think adding a bool would be more complicated-- member.Stop() would need to launch a goroutine to update a bool which has the same meaning as StopNotify() closed.

Comment (xiang90) [inline integration/cluster.go]: ok

**Comment by heyitsanthony** [comment]
ok all fixed ptal /c @xiang90

**Comment by xiang90** [comment]
LGTM if test passes

**Pull request merged**